src/dwn/models/record.js

```diff
@@ -0,0 +1,244 @@
+
+    // If a new `dateModified` was not provided, remove it from the updateMessage to let the DWN SDK auto-fill.
+    // This is necessary because otherwise DWN SDK throws an Error 409 Conflict due to attempting to overwrite a record
+    // when the `dateModified` timestamps are identical.
+    if (options?.dateModified === undefined) {
+      delete updateMessage.dateModified;
+    }
+
+    // If `published` is set to false, ensure that `datePublished` is undefined. Otherwise, DWN SDK's schema validation
+    // will throw an error if `published` is false but `datePublished` is set.
+    if (options?.published === false && updateMessage?.datePublished !== undefined) {
+      delete updateMessage.datePublished;
+    }
+
+    // Set the record ID and context ID, if any.
+    updateMessage.recordId = this.#recordId;
+    updateMessage.contextId = this.#contextId;
+
+    // Attempt to write the changes to mutable properties to the DWN.
+    const { message = null, record = null, status } = await this.#dwn.records.write(this.#target, {
+      author: this.#author,
+      data,
+      message: {
+        ...updateMessage,
+      },
+    });
+
+    if (status.code === 202 && record) {
+      // Only update the local Record instance mutable properties if the record was successfully (over)written.
+      mutableDescriptorProperties.forEach(property => {
+        this.#descriptor[property] = record[property];
+      });
+      // Only update data if `dataSize` is less than DWN 'max data size allowed to be encoded'.
+      if (data !== undefined) {
+        this.#readableStream = (record.dataSize <= DwnConstant.maxDataSizeAllowedToBeEncoded) ? record.data : null;
+        this.#encodedData = null; // Clear `encodedData` in case it was previously set.
+      }
+    }
+
+    return { message, status };
+  }
+
+  /**
+   * Utility methods.
+   */
+
+  /**
+   * Called by `JSON.stringify(...)` automatically.
+   */
+  toJSON() {
+    return {
+      author: this.author,
+      target: this.target,
+      recordId: this.id,
+      contextId: this.contextId,
+      dataFormat: this.dataFormat,
+      dateCreated: this.dateCreated,
+      interface: this.interface,
+      method: this.method,
+      parentId: this.parentId,
+      protocol: this.protocol,
+      recipient: this.recipient,
+      schema: this.schema,
+      dataCid: this.dataCid,
+      dataSize: this.dataSize,
+      dateModified: this.dateModified,
+      datePublished: this.datePublished,
+      published: this.published,
+    };
+  }
+
+  #freezeRecord() {
+    this.#isFrozen = true;
+  }
+
+  #unFreezeRecord() {
+    this.#isFrozen = false;
+  }
+
+  static #verifyPermittedMutation(propertiesToMutate, mutableDescriptorProperties) {
+    propertiesToMutate.forEach(propertyName => {
+      if (!mutableDescriptorProperties.includes(propertyName)) {
+        throw new Error(`${propertyName} is an immutable property. Its value cannot be changed.`);
+      }
+    });
+    return true;
```

Comment:
does this actually need to return anything?  it seems like it's just "success if it doesnt `throw`"

Reply:
You are correct that it doesn't need to return anything.

In an earlier iteration, I was re-using `verifyEqualityOfImmutableProperties()` from `dwn-sdk-js`:
https://github.com/TBD54566975/dwn-sdk-js/blob/07bcddb399d908fd431f4769dd0e48f1f03d9e68/src/interfaces/records/messages/records-write.ts#L553-L575

but ended up not wanting to have the same checks.  The `return true` stuck around after the refactor since it is in `verifyEqualityOfImmutableProperties()`.

We can remove the return but should likely update in both `web5-js` and `dwn-sdk-js` if we do.

Thoughts @dcrousso ?

Reply:
it feels a bit pointless to me (especially since you dont even check the return value when calling it above), so id personally remove it, but i dont feel super strongly

in a more general sense, i dont really see why we need to have exact matching of behavior/structure/etc. between entirely different repositories :/

Reply:
Addressed in https://github.com/TBD54566975/web5-js/commit/7a40a6622e2eb9fc885d89b83934f373ded6689f